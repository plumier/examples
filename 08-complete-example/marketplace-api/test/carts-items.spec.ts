import supertest from "supertest"

import createApp from "../src/app"
import { closeConnection, createShop, createUser, ignore, omitKeys } from "./_helper"

const omit = (obj: any) => omitKeys(obj, ["createdAt", "updatedAt"])

afterEach(async () => {
    await closeConnection()
})

describe("Charts Items", () => {
    it("Should add item properly", async () => {
        const app = await createApp({ mode: "production" })
        const { products } = await createShop(app, {
            items: [
                { name: "MacBook Pro 16 2020", basePrice: 2000, price: 2500 },
                { name: "Li-Ion Battery", basePrice: 200, price: 250 },
                { name: "M3 SSD 1TB", basePrice: 700, price: 900 },
            ]
        })
        const user = await createUser(app, { email: "jane.dane@gmail.com", name: "Jane Dane" })
        const { body: cart } = await supertest(app.callback())
            .get(`/api/carts/open`)
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        await supertest(app.callback())
            .post(`/api/carts/${cart.id}/items`)
            .send({ product: products[0].id, quantity: 1 })
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        await supertest(app.callback())
            .post(`/api/carts/${cart.id}/items`)
            .send({ product: products[1].id, quantity: 2 })
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        await supertest(app.callback())
            .post(`/api/carts/${cart.id}/items`)
            .send({ product: products[2].id, quantity: 2 })
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        const { body } = await supertest(app.callback())
            .get(`/api/carts/${cart.id}/items`)
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        expect(omit(body)).toMatchSnapshot()
    })
    it("Should merge item quantity when added twice", async () => {
        const app = await createApp({ mode: "production" })
        const { products } = await createShop(app, {
            items: [
                { name: "MacBook Pro 16 2020", basePrice: 2000, price: 2500 },
                { name: "Li-Ion Battery", basePrice: 200, price: 250 },
                { name: "M3 SSD 1TB", basePrice: 700, price: 900 },
            ]
        })
        const user = await createUser(app, { email: "jane.dane@gmail.com", name: "Jane Dane" })
        const { body: cart } = await supertest(app.callback())
            .get(`/api/carts/open`)
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        await supertest(app.callback())
            .post(`/api/carts/${cart.id}/items`)
            .send({ product: products[0].id, quantity: 3 })
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        await supertest(app.callback())
            .post(`/api/carts/${cart.id}/items`)
            .send({ product: products[0].id, quantity: 2 })
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        const { body } = await supertest(app.callback())
            .get(`/api/carts/${cart.id}/items`)
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        expect(omit(body)).toMatchSnapshot()
    })
    it("Should able to add from multiple shops", async () => {
        const app = await createApp({ mode: "production" })
        const { products: putraItems } = await createShop(app, {
            items: [
                { name: "MacBook Pro 16 2020", basePrice: 2000, price: 2500 },
                { name: "Li-Ion Battery", basePrice: 200, price: 250 },
                { name: "M3 SSD 1TB", basePrice: 700, price: 900 },
            ]
        })
        const { products: aliItems } = await createShop(app, {
            shop: { name: "Ali Shop" },
            owner: { email: "ali.owner@gmial.com" },
            staffs: [],
            items: [
                { name: "MacBook Pro 16 2020", basePrice: 2000, price: 2100 },
                { name: "Li-Ion Battery", basePrice: 200, price: 210 },
                { name: "M3 SSD 1TB", basePrice: 700, price: 800 },
            ]
        })
        const user = await createUser(app, { email: "jane.dane@gmail.com", name: "Jane Dane" })
        const { body: cart } = await supertest(app.callback())
            .get(`/api/carts/open`)
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        await supertest(app.callback())
            .post(`/api/carts/${cart.id}/items`)
            .send({ product: putraItems[0].id, quantity: 1 })
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        await supertest(app.callback())
            .post(`/api/carts/${cart.id}/items`)
            .send({ product: aliItems[0].id, quantity: 1 })
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        const { body } = await supertest(app.callback())
            .get(`/api/carts/${cart.id}/items`)
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        expect(omit(body)).toMatchSnapshot()
    })
    it("Should not accessible by other user", async () => {
        const app = await createApp({ mode: "production" })
        const jane = await createUser(app, { email: "jane.dane@gmail.com", name: "Jane Dane" })
        const john = await createUser(app, { email: "john.doe@ymail.com" })
        const { body: cart } = await supertest(app.callback())
            .get(`/api/carts/open`)
            .set("Authorization", `Bearer ${jane.token}`)
            .expect(200)
        await supertest(app.callback())
            .get(`/api/carts/${cart.id}/items`)
            .set("Authorization", `Bearer ${john.token}`)
            .expect(403)
    })
})