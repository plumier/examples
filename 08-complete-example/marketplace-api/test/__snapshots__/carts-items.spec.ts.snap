// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Charts Items Should able to add from multiple shops 1`] = `
Array [
  Object {
    "id": 1,
    "productId": 1,
    "productName": "MacBook Pro 16 2020",
    "productPrice": 2500,
    "quantity": 1,
    "shopId": 1,
    "shopName": "Putra Mahkota",
    "subTotal": 2500,
  },
  Object {
    "id": 2,
    "productId": 4,
    "productName": "MacBook Pro 16 2020",
    "productPrice": 2100,
    "quantity": 1,
    "shopId": 2,
    "shopName": "Ali Shop",
    "subTotal": 2100,
  },
]
`;

exports[`Charts Items Should add item properly 1`] = `
Array [
  Object {
    "id": 1,
    "productId": 1,
    "productName": "MacBook Pro 16 2020",
    "productPrice": 2500,
    "quantity": 1,
    "shopId": 1,
    "shopName": "Putra Mahkota",
    "subTotal": 2500,
  },
  Object {
    "id": 2,
    "productId": 2,
    "productName": "Li-Ion Battery",
    "productPrice": 250,
    "quantity": 2,
    "shopId": 1,
    "shopName": "Putra Mahkota",
    "subTotal": 500,
  },
  Object {
    "id": 3,
    "productId": 3,
    "productName": "M3 SSD 1TB",
    "productPrice": 900,
    "quantity": 2,
    "shopId": 1,
    "shopName": "Putra Mahkota",
    "subTotal": 1800,
  },
]
`;

exports[`Charts Items Should merge item quantity when added twice 1`] = `
Array [
  Object {
    "id": 1,
    "productId": 1,
    "productName": "MacBook Pro 16 2020",
    "productPrice": 2500,
    "quantity": 5,
    "shopId": 1,
    "shopName": "Putra Mahkota",
    "subTotal": 12500,
  },
]
`;
